### PR TITLE
More H3 header validation [#465].

### PR DESCRIPTION
**Pull request by rthalley**
This validates that:

* Content-Length should be validated against the actual size of the body

* Transfer-Encoding headers can only be accepted if the value is "trailers"

* Field names must not contain characters in the ranges 0x00-0x20, 0x41-0x5a, or 0x7f-0xff

* Except for pseudo-headers, headers must not contain single colons (:)

* Header values must not contain null bytes, \n (newlines) and \r (carriage returns)

* Header values must not contain \t (tabs) or spaces at the start or end

### CODE REVIEW

**Comment by rthalley** [comment]
Most of the checking was easy, but the content-length was a bit intrusive due to needing to track what we've seen, and the fact that between optimizations and the fact that streams can end in many ways, the required checks have to be done in several places.  Test coverage is still 100%.

File: src/aioquic/h3/connection.py [hunk 1]
```diff
@@ -194,19 +200,46 @@ def stream_is_request_response(stream_id: int):
     return stream_id % 4 == 0
 
 
+def validate_key(key: bytes):
```
Comment (jlaine) [inline src/aioquic/h3/connection.py:203]: validate_header_name might be a better name?

Also maybe explicitly type the return value to None and a docstring?

File: src/aioquic/h3/connection.py [hunk 2]
```diff
@@ -194,19 +200,46 @@ def stream_is_request_response(stream_id: int):
...
+
+
+def validate_value(key: bytes, value: bytes):
+    # RFC 9113 section 8.2.1
+    for c in value:
+        if c == NUL or c == LF or c == CR:
+            raise MessageError("Header %r value has forbidden characters" % key)
+    if len(value) > 0:
+        first = value[0]
+        if first == SP or first == TAB:
```
Comment (jlaine) [inline src/aioquic/h3/connection.py:221]: Are CR or LF whitespace too?

File: src/aioquic/h3/connection.py (hunk 1 above)
Comment (rthalley) [reply src/aioquic/h3/connection.py:203]: can do

File: src/aioquic/h3/connection.py (hunk 2 above)
Comment (rthalley) [reply src/aioquic/h3/connection.py:221]: CR and LF (along with NUL) are forbidden anywhere in the value by the RFC whereas SP and TAB are only banned as initial or final characters.  I contemplated making WHITESPACE(x) == (x == SP or x == TAB) but didn't as I in other situations WHITESPACE often includes CR and LF, and I wanted to make the translation from the RFC as literal as possible.  But I'm happy with making a WHITESPACE if you prefer it.

File: src/aioquic/h3/connection.py [hunk 3]
```diff
@@ -194,19 +200,46 @@ def stream_is_request_response(stream_id: int):
...
+    for i, c in enumerate(key):
+        if c <= 0x20 or (c >= 0x41 and c <= 0x5A) or c >= 0x7F:
+            raise MessageError("Header %r contains invalid characters" % key)
+        if c == COLON and i != 0:
+            # Colon not at start, definitely bad.  Keys starting with a colon
+            # will be checked in pseudo-header validation code.
+            raise MessageError("Header %r contains a non-initial colon" % key)
+
+
+def validate_value(key: bytes, value: bytes):
```
[the remark this answers is not in the document]
Comment (rthalley) [reply src/aioquic/h3/connection.py:214]: ok

File: src/aioquic/h3/connection.py [hunk 4]
```diff
@@ -230,6 +263,20 @@ def validate_headers(
...
+            if key == b"content-length":
+                try:
+                    content_length = int(value)
+                    if content_length < 0:
+                        raise ValueError
+                except ValueError:
+                    raise MessageError("content-length is not a non-negative integer")
+                if stream:
+                    stream.expected_content_length = content_length
+            elif key == b"transfer-encoding" and value != "trailers":
```
Comment (jlaine) [inline src/aioquic/h3/connection.py:276]: Shouldn't this be `b"trailers"` ? I'm not sure we have a test for the happy-path here :)

File: tests/test_h3.py [hunk 5]
```diff
@@ -1701,9 +1830,67 @@ def test_validate_request_headers(self):
...
+        )
+
+        # header with non-initial colon
+        with self.assertRaises(MessageError) as cm:
+            validate_request_headers([(b"x-f:oo", b"foo")])
+        self.assertEqual(
+            cm.exception.reason_phrase, "Header b'x-f:oo' contains a non-initial colon"
         )
 
+        # bad transfer-encoding
```
Comment (jlaine) [inline tests/test_h3.py:1859]: We're missing a test with a correct `transfer-encoding` :)

**Comment by jlaine** [comment]
This is great thanks!

File: src/aioquic/h3/connection.py (hunk 2 above)
Comment (jlaine) [reply src/aioquic/h3/connection.py:221]: I'd be fine with `WHITESPACE = (CR, LF) ` for instance, with a comment explaning why we don't need to test other sorts of whitespace.

**Comment by rthalley** [comment]
All feedback incorporated and the trailers bug fixed and tested both ways.  Will merge when the tests complete.
